packages/sign-client/src/controllers/engine.ts

```diff
@@ -369,11 +369,13 @@ export class Engine extends IEngine {
     this.events.once(engineEvent("session_update", clientRpcId), async ({ error }: any) => {
       if (error) reject(error);
       else {
-        await this.client.session.update(topic, { namespaces });
         resolve();
       }
     });
-
+    // update the session with the new namespaces, if the publish fails, revert to the old
+    // this allows the client to use the updated session like emitting events
+    // without waiting for the peer to acknowledge
```

Comment:
What's the maximum lag here between the client applying the update optimistically and then reverting it due to failed publish? Could see this looking like a weird "time travel" phenomenon, where the client acts on the optimistic update via side effects (like emitting events) and then reverts back to the pre-sessionUpdate state.

Reply:
reverting happens in maximum a minute if the relay doesn't acknowledge the publish. Once the publish is acked we can safely assume the dapp will receive and process the request as intended.
Additionally, with recent updates to [onSessionUpdateRequest](https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/sign-client/src/controllers/engine.ts#L1063) we can be sure the last sent update will be applied even when there are multiple received when the client is started.

>Could see this looking like a weird "time travel" phenomenon, where the client acts on the optimistic update via side effects (like emitting events) and then reverts back to the pre-sessionUpdate state.

technically yes, but imo it shouldn't be concern as clients are built to handle that case. For example this scenario
- wallet optimistically updates session with chain `x`
- wallet emits event with chain `x`
- publish timeouts and fails
- wallet reverts to session without chain 'x'

then when the dapp comes online it would receive event with chain 'x' but will discard it as it would fail validation https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/sign-client/src/controllers/engine.ts#L1264